[UI] Query logs using descending

peterschmidt85 opened the pull request:
Fixes #2892

This includes:

- [x] Allowing `descending` in pull log requests
- [x] Supporting `descending` in `FileLogStorage` (this implements reading of lines from a file in a reversed order)
- [x] Changing the pull log limit from 1000 to 100 (since now our chunks are larger than they are used to be)
- [x] Re-implemented to make `FileLogStorage._read_lines_reversed` fast regardless of `start_offset`

Plus, in addition to #2892, this PR updates `CloudWatchLogStorage` to implement retrying when it fetches empty log pages. The maximum number of retries is set to `10`. In case `ascending` is used, after hitting the limit `CloudWatchLogStorage` sets `next_token` to `None` (meaning no more logs). This is not ideal but given that ascending mode is not going to be used anymore, we can live with that for now. If `descending` is used, `CloudWatchLogStorage` returns `next_token` properly. (allowing the UI or CLI implement the logic of trying again).

Out of the scope:

- [ ] Updating `dstack log` to use `descending` and support `--start`

olgenn committed:
[UI] Query logs using descending and show timestamp per log entry #2892

peterschmidt85 committed:
[UI] Query logs using descending #2892

- [x] Allowed `descending` in `PollLogsRequest`
- [x] Supported `descending` in `FileLogStorage` (implemented an algorithm that reads lines from the end of t he file)
- [x] Updated `CloudWatchLogStorage` to support `MAX_RETRIES` (to skip empty pages)

peterschmidt85 commented:
Pending changes from @olgenn:

1. Fix the height of the `Logs` UI component.
2. Fix the background color of the `Logs` UI component.
3. Add `Legacy mode` toggler (off by default) that checks if log messages are base64-encoded strings and in this case decodes them.

olgenn committed:
[UI] Query logs using descending and show timestamp per log entry #2892

olgenn committed:
[UI] Query logs using descending and show timestamp per log entry #2892

olgenn committed:
[UI] Query logs using descending and show timestamp per log entry #2892

olgenn committed:
[UI] Query logs using descending and show timestamp per log entry #2892

olgenn committed:
[UI] Query logs using descending and show timestamp per log entry #2892

r4victor commented on src/dstack/_internal/server/services/logs/aws.py:
```diff
@@ -101,7 +103,45 @@ def poll_logs(self, project: ProjectModel, request: PollLogsRequest) -> JobSubmi
+                stream,
+                self.MAX_RETRIES,
+            )
+        return [], next_token if request.descending else None
```
Needs a comment explaining why returning next_token in case of descending and None otherwise

r4victor commented on src/dstack/_internal/server/services/logs/aws.py:
```diff
@@ -101,7 +103,45 @@ def poll_logs(self, project: ProjectModel, request: PollLogsRequest) -> JobSubmi
+                return [], None
+
+            if next_token == previous_next_token:
+                return [], None
```
seems natural to combine into one if

r4victor commented on src/dstack/_internal/server/services/logs/filelog.py:
```diff
@@ -94,6 +103,99 @@ def poll_logs(self, project: ProjectModel, request: PollLogsRequest) -> JobSubmi
+                raise ServerClientError(
+                    f"Invalid next_token for descending read: {request.next_token}. "
+                    f"Must be a non-negative integer offset."
+                )
```
Do we need different next_token parsing/validation for descending and ascending? Seems like it can be moved to a separate function.

r4victor commented on src/dstack/_internal/server/services/logs/filelog.py:
```diff
@@ -94,6 +103,99 @@ def poll_logs(self, project: ProjectModel, request: PollLogsRequest) -> JobSubmi
+            else:
+                f.seek(0, os.SEEK_END)
+                file_size = f.tell()
+                cursor = file_size
```
The two if branches are functionally the same.

r4victor commented:
> This is not ideal but given that ascending mode is not going to be used anymore, we can live with that for now.

You mean not used by the UI? Still used by the CLI and the API default t.

peterschmidt85 committed:
PR review feedback

peterschmidt85 committed:
PR review feedback (part 2)

peterschmidt85 committed:
PR review feedback (part 3)

peterschmidt85 committed:
PR review feedback (part 4)

peterschmidt85 committed:
PR review feedback (part 5)

Re-implemented to make `_read_lines_reversed` fast regardless of `start_offset`
